### PR TITLE
:squirrel: errors for no results

### PR DESCRIPTION
**Pull request by beseven**

### CODE REVIEW

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

File: test/unit/lib/messages.js [hunk 1]
```diff
@@ -60,4 +60,49 @@ describe('messages', () => {
...
+
+      expect(message).to.equal('<h2>Find your GP surgery</h2>  <p>You need to do this to go to your booking system. Enter the name of your surgery, ' +
+      'the name of your GP or a postcode.</p>');
+    });
+
+    it('should have a GP name message when only a name has been passed', () => {
+      const message = messages.noResultsMessage(res, true, undefined, 'netherthorpte');
+
+      expect(message).to.equal('<h2>We can not find a surgery using \'netherthorpte\'</h2> <p>Check the name you entered ' +
+        'is right. You get better results if you enter a full name. Or you can search with a postcode instead.</p>');
```
Comment (neilbmclaughlin) [inline test/unit/lib/messages.js]: I assume that this text will change in the review with Dean as it talks about 'Or ... postcode'

File: test/unit/lib/messages.js [hunk 2]
```diff
@@ -60,4 +60,49 @@ describe('messages', () => {
...
+    });
+
+    it('should have a combined message when an outcode and name have been passed', () => {
+      const message = messages.noResultsMessage(res, true, { isOutcode: true, term: 'TR21' }, 'Dave');
+
+      expect(message).to.equal('<h2>We can not find a surgery close to the \'TR21\' area and using \'Dave\'</h2> <p>Check the ' +
+        'location and name you entered are right and search again.</p>');
+    });
+
+    it('should have a combined error classe when a postcode and name have been passed', () => {
```
Comment (neilbmclaughlin) [inline test/unit/lib/messages.js]: Typo

File: app/views/includes/search-form.nunjucks [hunk 3]
```diff
@@ -10,19 +10,13 @@
       {% if errorMessage %}
         <h2>{{ searchErrorLabel }}</h2>
         <p>{{ errorMessage }}</p>
-      {% elif gps.length === 0 %}
-         <h2>We can't find a surgery using '{{search}}'</h2>
-         <p>Check the text you entered is right. You get better results if you enter a full name. Or you can search with a postcode instead.</p>
       {% else %}
-        <h2>Find your GP surgery</h2>
-        <p>You need to do this to go to your booking system. Enter the name of your surgery, the name of your GP or a postcode.</p>
+        {{ noResultsMessage | safe }}
```
Comment (neilbmclaughlin) [inline app/views/includes/search-form.nunjucks]: I felt OK using it in the previous work as there was no user input in the string being piped through safe. This is not the case here. I'm not sure if this introduces a vulnerability here. A quick check suggests not but I'd want to be sure.

File: app/lib/messages.js [hunk 4]
```diff
@@ -26,10 +27,35 @@ function searchInfomationMessage(singleResult, searchPostcode, searchTerm) {
...
-  return `${returnValue}.`;
+function searchInfomationMessage(singleResult, searchPostcode, searchTerm) {
+  return `We found ${(singleResult) ? 'this surgery' : 'these surgeries'}${promptBuilder(searchPostcode, searchTerm)}.`;
+}
+
+function noResultsMessage(res, noResult, searchPostcode, searchTerm) {
+  let headerPrompt = '';
+  let paragraphPrompt = '';
+  if (noResult) {
+    headerPrompt = `<h2>We can not find a surgery${promptBuilder(searchPostcode, searchTerm)}</h2>`;
```
Comment (neilbmclaughlin) [inline app/lib/messages.js]: I'm uncomfortable about this side effect. We shouldn't be returning the noResultsMessage and setting a res.locals value. An alternative would be to return an object (e.g. { header: string, paragraph: string, class: string })

File: app/views/includes/search-form.nunjucks (hunk 3 above)
Comment (beseven) [inline app/views/includes/search-form.nunjucks]: The pipe is there because of the h2 and p :)

File: test/unit/lib/messages.js (hunk 1 above)
Comment (beseven) [inline test/unit/lib/messages.js]: Talking to Dean about it now

File: app/views/includes/search-form.nunjucks (hunk 3 above)
Comment (neilbmclaughlin) [inline app/views/includes/search-form.nunjucks]: I know, same reason in the earlier work but is it a vector for an injection attack?

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

File: app/views/includes/search-form.nunjucks (hunk 3 above)
Comment (beseven) [inline app/views/includes/search-form.nunjucks]: Good point!

File: app/lib/messages.js (hunk 4 above)
Comment (beseven) [inline app/lib/messages.js]: You have a good point!

**Comment by beseven** [comment]
I have made Dean's content changes here, will rebase against the feature branch, use this URL for testing, and then work on making it mergeable // cc @neilbmclaughlin @ajthomascouk @ermlikeyeah

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

**Comment by neilbmclaughlin** [comment]
Just spotted that outcode value doesn't seem to being populated in the message:

http://gp-finder-pr-158.dev.beta.nhschoices.net/book-a-gp-appointment/results?search=&postcode=HG5

File: app/views/includes/search-form.nunjucks (hunk 3 above)
Comment (neilbmclaughlin) [inline app/views/includes/search-form.nunjucks]: An example to use would be searching <script>alert('Injected!');</script>. Google doesn't allow it to be posted and Safari just hangs. Searching by \<h1\>test\</h1\> _appears_ to sanitise the input.

**Comment by beseven** [comment]
@neilbmclaughlin, yeah - noticed that some of the merging went haywire :( a bit late. Will fix the glaring issues now.

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of gp-finder succeeded (http://gp-finder-pr-158.dev.beta.nhschoices.net)

**Pull request merged**